src/number_theory/legendre_symbol/gauss_sum.lean

```diff
@@ -0,0 +1,418 @@
+/-
+Copyright (c) 2022 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+import number_theory.legendre_symbol.add_character
+import number_theory.legendre_symbol.zmod_char
+import algebra.char_p.char_and_card
+
+/-!
+# Gauss sums
+
+We define the Gauss sum associated to a multiplicative and an additive
+character of a finite field and prove some results about them.
+
+## Main definition
+
+Let `R` be a finite commutative ring and let `R'` be another commutative ring.
+If `χ` is a multiplicative character `R → R'` (type `mul_char R R'`) and `ψ`
+is an additive character `R → R'` (type `add_char R R'`, which abbreviates
+`(multiplicative R) →* R'`), then the *Gauss sum* of `χ` and `ψ` is `∑ a, χ a * ψ a`.
+
+## Main results
+
+Some important results are as follows.
+
+* `char.gauss_sum_mul_gauss_sum_eq_card`: The product of the Gauss
+  sums of `χ` and `ψ` and that of `χ⁻¹` and `ψ⁻¹` is the cardinality
+  of the source ring `R` (if `χ` is nontrivial, `ψ` is primitive and `R` is a field).
+* `char.gauss_sum_sq`: The square of the Gauss sum is `χ(-1)` times
+  the cardinality of `R` if in addition `χ` is a quadratic character.
+* `char.quad_gauss_sum_frob`: For a quadratic character `χ`, raising
+  the Gauss sum to the `p`th power (where `p` is the characteristic of
+  the target ring `R'`) multiplies it by `χ p`.
+* `char.card_pow_card`: When `F` and `F'` are finite fields and `χ : F → F'`
+  is a nontrivial quadratic character, then `(χ (-1) * #F)^(#F'/2) = χ (#F')`.
+* `char.two_pow_card`: For every finite field `F` of odd characteristic,
+  we have `2^(#F/2) = χ₈(#F)` in `F`.
+
+This machinery can be used to derive (a generaliation of) the Law of
+Quadratic Reciprocity.
+
+## Tags
+
+additive character, multiplicative character, Gauss sum
+-/
+
+namespace char
```

Comment:
Why do you put everything in the `char` namespace? Is that used in other places as well?
I understand that it doesn't really make sense to choose `mul_char` xor `add_char`. But almost all the names contain `gauss_sum_` as prefix, so those could go in the root namespace.
The remaining names can maybe be placed in an appropriate namespace manually.

Reply:
I'm just hesitant to put stuff in the root namespace. I'm now putting everything starting with "gauss_sum_" into the root namespace; then there are `char.card_pow_char_pow`, `char.card_pow_card` and `finite_field.two_pow_card`.
(The `char` namespace is used for stuff on quadratic characters (number_theory/legendre_symbol/quadratic_char.lean), but this is perhaps also up for debate, as most names there involve `quadratic_char_`.)

Reply:
Ok, that seems good to me for now. We can tweak the namespaces further as more material is added and the organization of the lemma names crystalizes.